File: src/main/java/com/github/bin/service/HisMsgService.java
package com.github.bin.service;

import com.github.bin.command.Command;
import com.github.bin.config.MsgDataSource;
import com.github.bin.mapper.msg.HisMsgMapper;
import lombok.extern.slf4j.Slf4j;
import lombok.val;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.function.Consumer;
import java.util.function.Function;

/**
 * @author bin
 * @since 2023/08/22
 */
@Service
@Slf4j
public class HisMsgService {
    private static HisMsgMapper hisMsgMapper;
    private static List<Command> commands;

    @Autowired
    public void setHisMsgMapper(HisMsgMapper hisMsgMapper) {
        HisMsgService.hisMsgMapper = hisMsgMapper;
    }

    @Autowired
    public void setCommands(List<Command> commands) {
        HisMsgService.commands = commands;
    }

    public static <T> T apply(String tableName, Function<HisMsgMapper, T> block) {
        MsgDataSource.set(tableName);
        return block.apply(hisMsgMapper);
    }

    public static void accept(String tableName, Consumer<HisMsgMapper> block) {
        MsgDataSource.set(tableName);
        block.accept(hisMsgMapper);
    }

    public static void handleBot(RoomConfig roomConfig, String id, String msg) {
        try {
            for (val command : commands) {
                if (command.invoke(roomConfig, id, msg)) {
                    return;
                }
            }
        } catch (Exception e) {
            log.warn("bot 处理异常", e);
            roomConfig.sendAsBot("【ERROR】bot 处理异常，请联系管理员");
        }
    }

    public static void sendAsBot(RoomConfig roomConfig, String msg) {
        roomConfig.sendAsBot(msg);
    }
}

File: src/main/kotlin/com/github/bin/service/HisMsgService.kt
package com.github.bin.service

import com.github.bin.command.Command
import com.github.bin.config.MsgDataSource
import com.github.bin.mapper.msg.HisMsgMapper
import com.github.bin.model.Message
import org.springframework.stereotype.Service

/**
 * @author bin
 * @version 1.0.0
 * @since 2023/7/9
 */
@Suppress("LeakingThis")
@Service
class HisMsgService(
    private val hisMsgMapper: HisMsgMapper,
    private val commands: List<Command>
) {
    init {
        INSTANCE = this
    }

    fun <T> invoke(tableName: String, block: HisMsgMapper.() -> T): T {
        MsgDataSource.set(tableName)
        return hisMsgMapper.block()
    }

    fun handleBot(roomConfig: RoomConfig, id: String, msg: String) {
        for (command in commands) {
            if (command.invoke(roomConfig, id, msg)) {
                break
            }
        }
    }

    companion object {
        private lateinit var INSTANCE: HisMsgService
        fun RoomConfig.sendAsBot(msg: String) {
            val text = Message.Text()
            text.msg = msg
            sendAsBot(text)
        }

        private fun RoomConfig.sendAsBot(msg: Message.Msg) {
            msg.role = -10
            INSTANCE.invoke(id) {
                msg.id = insert(msg.type, msg.msg, -10)
            }
            sendAll(msg)
        }
    }
}
